really-core/src/main/scala/io/really/gorilla/Replayer.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (C) 2014-2015 Really Inc. <http://really.io>
+ */
+
+package io.really.gorilla
+
+import akka.actor._
+import akka.contrib.pattern.DistributedPubSubMediator.Unsubscribe
+import akka.util.Timeout
+import io.really.gorilla.Replayer.SnapshotResult._
+import io.really.gorilla.Replayer.{ EndOfEventsLog, SnapshotResult }
+import io.really.model.CollectionActor.{ GetState, State }
+import io.really.{ R, Revision, ReallyGlobals }
+import akka.pattern.{ AskTimeoutException, ask, pipe }
+import play.api.libs.json.JsObject
+import io.really.CommandError.ObjectNotFound
+import io.really.protocol.SubscriptionFailure
+
+import scala.concurrent.Future
+import scala.slick.lifted.TableQuery
+import scala.slick.driver.H2Driver.simple._
+import scala.util.control.NonFatal
+import scala.concurrent.ExecutionContext.Implicits.global
+import EventLogs._
+
+class Replayer(globals: ReallyGlobals, objectSubscriber: ActorRef, rSubscription: RSubscription,
+    maxMarker: Option[Revision])(implicit session: Session) extends Actor with ActorLogging with Stash {
+
+  val replayerId = s"Replayer ${rSubscription.pushChannel.path}$$${rSubscription.r}"
+
+  val min = rSubscription.rev
+  private[gorilla] val r = rSubscription.r
+
+  context.watch(objectSubscriber)
+  //TODO Watch Gorilla to notify the ObjectSubscriber that the subscription failed in case of Gorilla's termination
+
+  maxMarker.map {
+    max: Revision =>
+      if (min == max) {
+        context.become(servePushUpdates)
+      } else if (min > max && (min - max) <= globals.config.GorillaConfig.advancedRevisionLimit) {
+        context.become(waitingCorrectRevision)
+      } else if (min > max) {
+        shutdownWithObjectSubscriptionAck("Subscription revision advances the current known revision with more than" +
+          " the advancing limit")
+      } else {
+        getEvents(r, min)
+      }
+  }.getOrElse {
+    getEvents(r, min)
+  }
+
+  def commonHandler: Receive = {
+    case SubscriptionManager.Unsubscribe =>
+      shutdown("Received unsubscribe request")
+    case Terminated(`objectSubscriber`) =>
+      shutdown("Object Subscription Actor was stopped")
+    case msg =>
+      unexpectedMessage(msg.toString)
+  }
+
+  def receive: Receive = starterReceiver orElse commonHandler
+
+  def starterReceiver: Receive = {
+    case _: GorillaLogEntry =>
+      stash()
+    case SnapshotObject(obj) =>
+      //TODO State should have a revision field
+      val stateRev = (obj \ "_rev").as[Long]
+      if (min > stateRev && (min - stateRev) <= globals.config.GorillaConfig.advancedRevisionLimit) {
+        context.become(waitingCorrectRevision)
+      } else if (min > stateRev) {
+        shutdownWithObjectSubscriptionAck("Subscription revision advances the current known revision with more than" +
+          " the advancing limit")
+      } else {
+        objectSubscriber ! SnapshotObject(obj)
+        context.become(servePushUpdates)
+      }
+    case SnapshotFetchError(r, _) =>
```

Comment:
no need to make this error take `r` and reason, this error should be `case object`